src/beast/beast/asio/append_buffers.h

```diff
@@ -0,0 +1,516 @@
+    append_buffers_helper(append_buffers_helper const&) = default;
+    append_buffers_helper& operator=(append_buffers_helper&&) = default;
+    append_buffers_helper& operator=(append_buffers_helper const&) = default;
+
+    explicit
+    append_buffers_helper(Bs const&... bs)
+        : bs_(bs...)
+    {
+    }
+
+    const_iterator
+    begin() const;
+
+    const_iterator
+    end() const;
+};
+
+template<class ValueType, class... Bs>
+class append_buffers_helper<
+    ValueType, Bs...>::const_iterator
+{
+    template<class U>
+    static
+    inline
+    std::size_t constexpr
+    max_sizeof()
+    {
+        return sizeof(U);
+    }
+
+    template<class U0, class U1, class... Us>
+    static
+    inline
+    std::size_t constexpr
+    max_sizeof()
+    {
+        return
+            max_sizeof<U0>() > max_sizeof<U1, Us...>() ?
+            max_sizeof<U0>() : max_sizeof<U1, Us...>();
+    }
+
+    std::size_t n_;
+    std::tuple<Bs...> const* bs_;
+    std::array<std::uint8_t,
+        max_sizeof<typename Bs::const_iterator...>()> buf_;
+
+    friend class append_buffers_helper<ValueType, Bs...>;
+
+    template<std::size_t I>
+    using C = std::integral_constant<std::size_t, I>;
+
+    template<std::size_t I>
+    using iter_t = typename std::tuple_element_t<
+        I, std::tuple<Bs...>>::const_iterator;
+
+    template<std::size_t I>
+    inline
+    iter_t<I>&
+    iter()
+    {
+        return *reinterpret_cast<
+            iter_t<I>*>(buf_.data());
+    }
+
+    template<std::size_t I>
+    inline
+    iter_t<I> const&
+    iter() const
+    {
+        return *reinterpret_cast<
+            iter_t<I> const*>(buf_.data());
+    }
+
+public:
+    using value_type = ValueType;
+    using pointer = value_type const*;
+    using reference = value_type;
+    using difference_type = std::ptrdiff_t;
+    using iterator_category =
+        std::bidirectional_iterator_tag;
+
+    ~const_iterator();
+    const_iterator();
+    const_iterator(const_iterator&& other);
+    const_iterator(const_iterator const& other);
+    const_iterator& operator=(const_iterator&& other);
+    const_iterator& operator=(const_iterator const& other);
+
+    bool
+    operator==(const_iterator const& other) const;
+
+    bool
+    operator!=(const_iterator const& other) const
+    {
+        return !(*this == other);
+    }
+
+    reference
+    operator*() const;
+
+    pointer
+    operator->() const = delete;
+
+    const_iterator&
+    operator++();
+
+    const_iterator
+    operator++(int)
+    {
+        auto temp = *this;
+        ++(*this);
+        return temp;
+    }
+
+    const_iterator&
+    operator--();
+
+    const_iterator
+    operator--(int)
+    {
+        auto temp = *this;
+        --(*this);
+        return temp;
+    }
+
+private:
+    const_iterator(
+        std::tuple<Bs...> const& bs, bool at_end);
+
+    inline
+    void
+    construct(C<sizeof...(Bs)>)
+    {
+        auto constexpr I = sizeof...(Bs);
+        n_ = I;
```

Comment:
Its intentional, so that the body of the function reads consistently with the other functions (which refer to `I`)